File: sahayog_asset/sahayog_stock/report/item_list/item_list.js
// Copyright (c) 2024, Sid and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Item List"] = {
  filters: [
    {
      fieldname: "status",
      label: __("Status"),
      fieldtype: "Select",
      options: ["", "Low Stock", "Out of Stock", "In Stock"],
      default: "",
      reqd: 0,
    },

    {
      label: "Add New Item",
      fieldname: "add_new",
      fieldtype: "Button",
      width: 200,
    },

    {
      label: "Create PR for Low Stock",
      fieldname: "create_pr_low_stock",
      fieldtype: "Button",
      width: 200,
    },

    {
      label: "Create PR for Out of Stock",
      fieldname: "create_pr_out_of_stock",
      fieldtype: "Button",
      width: 200,
    },
    {
      label: "Stock Entry",
      fieldname: "stock_entry",
      fieldtype: "Button",
      width: 200,
    },
    {
      label: "Purchase Requisition",
      fieldname: "p_r",
      fieldtype: "Button",
      width: 200,
    },
  ],

  formatter: function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);

    if (column.fieldname == "status") {
      let color = "black";
      let backgroundColor = "transparent";

      if (value == "Out of Stock") {
        color = "red";
        backgroundColor = "rgba(255, 0, 0, 0.1)"; // Faded red
      } else if (value == "Low Stock") {
        color = "orange";
        backgroundColor = "rgba(255, 165, 0, 0.1)"; // Faded orange
      } else if (value == "In Stock") {
        color = "green";
        backgroundColor = "rgba(0, 128, 0, 0.1)"; // Faded green
      }

      value = `<b style="color:${color}; background-color:${backgroundColor};">${value}</b>`;
    }

    return value;
  },
};

$(document).on("click", "button[data-fieldname='stock_entry']", function () {
  const route = "/app/store-entry/view/list?category=IT";
  frappe.set_route(route);
});

$(document).on("click", "button[data-fieldname='add_new']", function () {
  const route = "/app/sahayog-item/new-sahayog-item";
  frappe.set_route(route);
});
$(document).on("click", "button[data-fieldname='p_r']", function () {
  const route = "/app/purchase-requisition";
  frappe.set_route(route);
});
// Open custom dialog box on button click
// Open custom dialog box on button click
$(document).on(
  "click",
  "button[data-fieldname='create_pr_low_stock']",
  function () {
    // Fetch low stock items from the server-side method
    frappe.call({
      method: "sahayog_asset.automation.get_low_stock_items",
      callback: function (response) {
        // Extract the low stock items from the response
        const lowStockItems = response.message || [];

        // Create and show the custom dialog
        let d = new frappe.ui.Dialog({
          title: "Low Stock Items",
          fields: [
            {
              label: "Low Stock Items",
              fieldname: "low_stock_items",
              fieldtype: "HTML",
              options: `
                          <table class="table table-bordered">
                              <thead>
                                  <tr>
                                      <th>Item Code</th>
                                      <th>Item Name</th>
                                      <th>New Quantity</th>
                                  </tr>
                              </thead>
                              <tbody id="low-stock-table-body">
                                  ${lowStockItems
                                    .map(
                                      (item) => `
                                              <tr>
                                                  <td>${item.name}</td>
                                                  <td>${item.item_name}</td>
                                                  <td>
                                                      <input type="number" 
                                                             class="form-control" 
                                                             data-item-code="${item.name}" 
                                                             data-item-name="${item.item_name}" 
                                                             placeholder="Enter new quantity" 
                                                             min="1">
                                                  </td>
                                              </tr>
                                          `
                                    )
                                    .join("")}
                              </tbody>
                          </table>
                      `,
            },
          ],
          size: "small",
          primary_action_label: "Create Purchase Requisition",
          primary_action(values) {
            // Collect user inputs and process them
            const inputs = Array.from(
              document.querySelectorAll("#low-stock-table-body input")
            );
            const updatedQuantities = inputs
              .map((input) => ({
                item_code: input.getAttribute("data-item-code"),
                item_name: input.getAttribute("data-item-name"),
                new_quantity: parseInt(input.value, 10) || 1, // Default to 1 if empty or zero
              }))
              .filter((item) => item.new_quantity > 0);

            if (updatedQuantities.length === 0) {
              frappe.msgprint({
                title: "No Valid Quantities",
                message:
                  "No valid quantities entered. Please enter quantities greater than zero.",
                indicator: "red",
              });
              return;
            }

            // Send updatedQuantities to the server for processing
            frappe.call({
              method: "sahayog_asset.automation.create_purchase_requisition",
              args: {
                quantities: JSON.stringify(updatedQuantities),
              },
              callback: function (response) {
                if (response.message.error) {
                  frappe.msgprint({
                    title: "Error",
                    message: response.message.error,
                    indicator: "red",
                  });
                } else {
                  frappe.show_alert(
                    {
                      message: __("Purchase Requisition created successfully"),
                      indicator: "green",
                    },
                    5
                  );
                  frappe.set_route("List", "Purchase Requisition", {
                    status: "Draft",
                  });

                  d.hide();
                }
              },
            });
          },
          secondary_action_label: "Cancel",
          secondary_action() {
            d.hide();
          },
        });

        d.show();
      },
    });
  }
);

$(document).on(
  "click",
  "button[data-fieldname='create_pr_out_of_stock']",
  function () {
    // Fetch out of stock items from the server-side method
    frappe.call({
      method: "sahayog_asset.automation.get_out_of_stock_items",
      callback: function (response) {
        // Extract the out of stock items from the response
        const lowStockItems = response.message || [];

        // Create and show the custom dialog
        let d = new frappe.ui.Dialog({
          title: "Out of Stock Items",
          fields: [
            {
              label: "Out of Stock Items",
              fieldname: "out_of_stock_items",
              fieldtype: "HTML",
              options: `
                          <table class="table table-bordered">
                              <thead>
                                  <tr>
                                      <th>Item Code</th>
                                      <th>Item Name</th>
                                      <th>New Quantity</th>
                                  </tr>
                              </thead>
                              <tbody id="out-of-stock-table-body">
                                  ${lowStockItems
                                    .map(
                                      (item) => `
                                              <tr>
                                                  <td>${item.name}</td>
                                                  <td>${item.item_name}</td>
                                                  <td>
                                                      <input type="number" 
                                                             class="form-control" 
                                                             data-item-code="${item.name}" 
                                                             data-item-name="${item.item_name}" 
                                                             placeholder="Enter new quantity" 
                                                             min="1">
                                                  </td>
                                              </tr>
                                          `
                                    )
                                    .join("")}
                              </tbody>
                          </table>
                      `,
            },
          ],
          size: "small",
          primary_action_label: "Create Purchase Requisition",
          primary_action(values) {
            // Collect user inputs and process them
            const inputs = Array.from(
              document.querySelectorAll("#out-of-stock-table-body input")
            );
            const updatedQuantities = inputs
              .map((input) => ({
                item_code: input.getAttribute("data-item-code"),
                item_name: input.getAttribute("data-item-name"),
                new_quantity: parseInt(input.value, 10) || 1, // Default to 1 if empty or zero
              }))
              .filter((item) => item.new_quantity > 0);

            if (updatedQuantities.length === 0) {
              frappe.msgprint({
                title: "No Valid Quantities",
                message:
                  "No valid quantities entered. Please enter quantities greater than zero.",
                indicator: "red",
              });
              return;
            }

            // Send updatedQuantities to the server for processing
            frappe.call({
              method: "sahayog_asset.automation.create_purchase_requisition",
              args: {
                quantities: JSON.stringify(updatedQuantities),
              },
              callback: function (response) {
                if (response.message.error) {
                  frappe.msgprint({
                    title: "Error",
                    message: response.message.error,
                    indicator: "red",
                  });
                } else {
                  frappe.msgprint({
                    title: "Purchase Requisition Created",
                    message: `Purchase Requisition creation has been queued.`,
                    indicator: "green",
                  });

                  frappe.set_route("List", "Purchase Requisition", {
                    status: "Draft",
                  });

                  d.hide();
                }
              },
            });
          },
          secondary_action_label: "Cancel",
          secondary_action() {
            d.hide();
          },
        });

        d.show();
      },
    });
  }
);

// $(document).ready(function () {
//   // Apply background color, text styling, and responsive width using jQuery
//   $("button[data-fieldname='add_new']").css({
//     "background-color": "black", // Background color
//     color: "white", // Text color
//     "border-radius": "7px", // Rounded corners
//     "font-weight": "bold", // Bold text
//   });
// });

// $(document).ready(function () {
//   // Apply background color, text styling, and responsive width using jQuery
//   $("button[data-fieldname='create_pr_low_stock']").css({
//     "background-color": "black", // Background color
//     color: "white", // Text color
//     "border-radius": "7px", // Rounded corners
//     "font-weight": "bold", // Bold text
//   });
// });

// $(document).ready(function () {
//   // Apply background color, text styling, and responsive width using jQuery
//   $("button[data-fieldname='create_pr_out_of_stock']").css({
//     "background-color": "black", // Background color
//     color: "white", // Text color
//     "border-radius": "7px", // Rounded corners
//     "font-weight": "bold", // Bold text
//   });
// });
// $(document).ready(function () {
//   // Apply background color, text styling, and responsive width using jQuery
//   $("button[data-fieldname='stock_entry']").css({
//     "background-color": "black", // Background color
//     color: "white", // Text color
//     "border-radius": "7px", // Rounded corners
//     "font-weight": "bold", // Bold text
//   });
// });

File: sahayog_asset/sahayog_stock/report/item_list/item_list.py
import frappe

def execute(filters=None):
    if not filters:
        filters = {}

    # Determine user role and set category filter
    user_roles = frappe.get_roles(frappe.session.user)
    
    if "System Manager" in user_roles:
        category_filter = None  # No filter for System Manager
    elif "IT Store Manager" in user_roles or "IT Support Manager" in user_roles:
        category_filter = "IT"
    elif "Stationery Asset Admin" in user_roles:
        category_filter = "Stationery"
    elif "Admin Store Manager" in user_roles:
        category_filter = "Admin"
    else:
        category_filter = None  # Default category if none of the roles match

    status_filter = filters.get("status")
    columns = get_columns()
    data = get_cs_data(category_filter, status_filter)
    
    # Calculate the report summary
    in_stock_count = sum(1 for row in data if row["status"] == "In Stock")
    low_stock_count = sum(1 for row in data if row["status"] == "Low Stock")
    out_of_stock_count = sum(1 for row in data if row["status"] == "Out of Stock")
    
    report_summary = [
        {"label": "In Stock", "value": in_stock_count, "indicator": "green"},
        {"label": "Low Stock", "value": low_stock_count, "indicator": "red"},
        {"label": "Out of Stock", "value": out_of_stock_count, "indicator": "red"}
    ]

    # Return columns, data, message, chart, and report summary
    return columns, data, None, None, report_summary

def get_columns():
    return [
        {
            "fieldname": "status",
            "label": '<b style="color:black">Status</b>',
            "fieldtype": "Data",
            "width": "180",
        },
        {
            "fieldname": "name",
            "label": '<b style="color:black">ID</b>',
            "fieldtype": "Link",
            "options": "Sahayog Item",
            "width": "180",
        },
        {
            "fieldname": "item_name",
            "label": '<b style="color:black">Item Name</b>',
            "fieldtype": "Data",
            "width": "700",
        },
        {
            "fieldname": "current_stock",
            "label": '<b style="color:black">Current Stock</b>',
            "fieldtype": "Int",
            "width": "150",
        },
        {
            "fieldname": "min_qty",
            "label": '<b style="color:black">Minimum Qty</b>',
            "fieldtype": "Int",
            "width": "150",
        },
        {
            "fieldname": "category",
            "label": '<b style="color:black">Category</b>',
            "fieldtype": "Data",
            "width": "100",
        },
    ]

def get_cs_data(category_filter, status_filter):
    # Build the SQL query for detailed data
    sql_query = """
    SELECT
        si.name, si.item_name, si.current_stock, si.min_qty, si.category, si.modified,
        CASE
            WHEN si.current_stock <= 0 THEN 'Out of Stock'
            WHEN si.current_stock <= si.min_qty THEN 'Low Stock'
            ELSE 'In Stock'
        END AS status
    FROM
        `tabSahayog Item` AS si
    WHERE
        (%s IS NULL OR si.category = %s)
        AND (%s IS NULL OR
            (CASE
                WHEN si.current_stock <= 0 THEN 'Out of Stock'
                WHEN si.current_stock <= si.min_qty THEN 'Low Stock'
                ELSE 'In Stock'
            END) = %s
        )
   ORDER BY modified DESC
    """
    
    # Execute the SQL query with category and status filters
    data = frappe.db.sql(sql_query, (category_filter, category_filter, status_filter, status_filter), as_dict=True)

    return data
